Readme.md

# cruise-db

  A reference k/v store based upon [cruise][cruise]. It's the javascript analog to [`goraft/raftd`][raftd].

[cruise]: https://github.com/calvinfo/cruise
[raftd]: https://github.com/goraft/raftd

## Usage

  To get a server running you'll want to run the server:


    $ npm install -g cruise-db
    $ cruise-db --port 4001
    listening on 4001...

  Then you can store values by posting to `/db/:key`

    $ curl http://localhost:4001/db/some-key \
        -X POST \
        --data 'foo=bar'

  And get them back from `/db/:key`

    $ curl http://localhost:4001/db/some-key
    {"foo":"bar"}

## Clustering

  Since the whole point of raft is consensus among a distributed system, the reference db can be run as a cluster. Simply pass the `--join [-j]` option with the connection string of the leader to add servers to the cluster

    $ cruise-db
    Listening on 4001...

    $ cruise-db --port 4002 --join localhost:4001
    Listening on 4002...

    $ cruise-db --port 4003 --join localhost:4001
    Listening on 4003...

  You can query keys from the followers and they should stay in sync with any new keys posted to the leader.


## Caveats

  Just like raftd, the system does not do any sort of client forwarding. In order to submit new commands, you will have to post them to the leader node.

## License

(The MIT License)

Copyright (c) 2014 Calvin French-Owen &lt;calvin@calv.info&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

History.md

0.0.2 / 2014-05-25
==================

 * splitting tests, updating with api fixes

0.0.1 / 2014-05-17
==================

  * Initial release


test/test.js
var Server = require('../');
var test = require('supertest');
var assert = require('assert');
var after = require('after');

describe('simple cruise-db', function(){
  var app;
  var server;
  beforeEach(function(done){
    server = new Server();
    app = server.app();
    var interval = setInterval(function(){
      if (!server.cruise.isLeader()) return;
      clearInterval(interval);
      done();
    }, 100);
  });

  describe('/join', function(){
    it('should return the host and port of the cruise leader', function(done){
      var cruise = server.cruise;
      test(app)
        .post('/join')
        .send({ addr: '127.0.0.1:4030' })
        .expect({
          peers: ['127.0.0.1:4030', cruise.addr()]
        })
        .end(function(err){
          assert(!err);
          assert.equal(cruise.peers().length, 1);
          done();
        });
    });
  });

  describe('/put', function(){
    it('should return filled keys', function(done){
      var key = 'key';
      var val = { foo: 'bar' };
      test(app)
        .post('/db/' + key)
        .send(val)
        .expect(200)
        .end(function(){
          test(app)
            .get('/db/' + key)
            .expect(200)
            .expect(val)
            .end(done);
        });
    });
  });

  describe('/get', function(){
    it('should return 404 for empty keys', function(done){
      test(app)
        .get('/db/non-existent')
        .expect(404)
        .end(done);
    });
  });
});

describe('clustering cruise-db', function(){
  var servers = [];

  before(function(done){
    for (var i = 0; i < 3; i++) servers.push(new Server());
    var nodes = servers.map(function(server){ return server.cruise; });
    nodes.forEach(function(node){
      nodes.forEach(function(peer){
        node.addPeer(peer.addr());
      });
    });

    setTimeout(done, 1000);
  });

  it('should record a key from the master', function(done){
    var leader = servers[0];
    done = after(3, done);
    test(leader.app())
      .post('/db/key')
      .send({ foo: 'bar' })
      .expect(200)
      .end(function(err){
        assert(!err);
        setTimeout(verify, 300);
      });

    function verify(){
      servers.forEach(function(server){
        test(server.app())
          .get('/db/key')
          .expect({ foo: 'bar' })
          .end(done);
      });
    }
  });
});

lib/index.js
var bindAll = require('bind-all');
var bodyParser = require('body-parser');
var Cruise = require('cruise');
var Db = require('./db');
var debug = require('debug')('cruise-db:server');
var express = require('express');

/**
 * Module `exports`.
 */

module.exports = Server;

/**
 * Server constructor
 */

function Server(opts){
  if (!this instanceof Server) return new Server(opts);
  opts = opts || {};
  var addr = opts.addr || '127.0.0.1:' + port();
  this.cruise = new Cruise(addr);
  this.cruise.listen();
  var db = this.db = new Db();
  this.cruise
    .on('data', function(data){ db.put(data.key, data.val); })
    .on('add peer', function(peer){ debug('added peer %s', peer); });
  bindAll(this);
}

/**
 * Creates a new app for mounting
 *
 * @return {App} app
 */

Server.prototype.app = function(){
  var app = express()
    .use(bodyParser())
    .post('/join', this.join)
    .post('/db/:key', this.put)
    .get('/db/:key', this.get);
  return app;
};

/**
 * Joins a peer to the leader node
 *
 * @param {Request} req
 *   @param {String} addr  e.g. 127.0.0.1:4001
 * @param {Response} res
 * @param {Function} next
 */

Server.prototype.join = function(req, res, next){
  var addr = req.param('addr');
  var cruise = this.cruise;
  cruise.join(addr);
  var peers = cruise
    .peers()
    .map(function(peer){
      return peer.addr();
    });
  peers.push(cruise.addr());
  res.send(200, {
    peers: peers
  });
};

/**
 * Set a value in our database
 *
 * @param {Request} req
 *   @param {String} key
 *   @param {Mixed} body  the value to set
 * @param {Response} res
 * @param {Function} next
 */

Server.prototype.put = function(req, res, next){
  var key = req.param('key');
  var val = req.body;
  debug('recording %s: %j', key, val);
  this.cruise.do({ key: key, val: val }, function(err){
    if (err) return res.send(500, err.message);
    res.send(200);
  });
};

/**
 * Retrieves the value from the database
 *
 * @param {Request} req
 *   @param {String} key
 * @param {Response} res
 * @param {Function} next
 */

Server.prototype.get = function(req, res, next){
  var key = req.param('key');
  var val = this.db.get(key);
  if (val == null) return res.send(404);
  res.send(val);
};

/**
 * Get a random port for the cruise server
 */

function port(){
  return Math.floor((Math.random() * 55536) + 10000);
}